pkg/scheduler/framework/runtime/batch.go

```diff
@@ -0,0 +1,236 @@
+}
+
+// OpportunisticBatching caches results from filtering and scoring when possible to optimize
+// scheduling of common pods.
+type OpportunisticBatch struct {
+	state         *batchState
+	lastCycle     schedulingCycle
+	signatureFunc PodSignatureFunc
+	handle        fwk.Handle
+
+	// Used primarily for tests, count the total pods we
+	// have successfully batched.
+	batchedPods int64
+}
+
+type batchState struct {
+	signature    string
+	sortedNodes  framework.SortedScoredNodes
+	creationTime time.Time
+}
+
+type schedulingCycle struct {
+	cycleCount int64
+	chosenNode string
+	succeeded  bool
+}
+
+const (
+	maxBatchAge = 500 * time.Millisecond
+)
+
+// Provide a hint for the pod based on filtering an scoring results of previous cycles. Caching works only for consecutive pods
+// with the same signature that are scheduled in 1-pod-per-node manner (otherwise previous scores could not be reused).
+// It's assured by checking the top rated node is no longer feasible (meaning the previous pod was successfully scheduled and the
+// current one does not fit).
+func (b *OpportunisticBatch) GetNodeHint(ctx context.Context, pod *v1.Pod, state fwk.CycleState, cycleCount int64) (string, string) {
+	logger := klog.FromContext(ctx)
+	var hint string
+
+	startTime := time.Now()
+	defer func() {
+		hinted := "hint"
+		if hint == "" {
+			hinted = "no_hint"
+		}
+		metrics.FrameworkExtensionPointDuration.WithLabelValues(metrics.GetNodeHint, hinted, b.handle.ProfileName()).Observe(metrics.SinceInSeconds(startTime))
+	}()
+
+	signature := b.signatureFunc(b.handle, ctx, pod)
+
+	nodeInfos := b.handle.SnapshotSharedLister().NodeInfos()
+
+	// If we don't have state that we can use, then return an empty hint.
+	if !b.batchStateCompatible(ctx, logger, pod, signature, cycleCount, state, nodeInfos) {
+		metrics.BatchAttemptStats.WithLabelValues(b.handle.ProfileName(), metrics.BatchAttemptNoHint).Inc()
+		logger.V(4).Info("OpportunisticBatch no node hint available for pod",
+			"profile", b.handle.ProfileName(), "pod", klog.KObj(pod), "cycleCount", cycleCount)
+		return "", signature
+	}
+
+	// Otherwise, pop the head of the list in our state and return it as
+	// a hint. Also record it in our data to compare on storage.
+	hint = b.state.sortedNodes.Pop()
+	logger.V(3).Info("OpportunisticBatch provided node hint",
+		"profile", b.handle.ProfileName(), "pod", klog.KObj(pod), "cycleCount", cycleCount, "hint", hint,
+		"remainingNodes", b.state.sortedNodes.Len())
+
+	return hint, signature
+}
+
+// Store results from scheduling for use as a batch later.
+func (b *OpportunisticBatch) StoreScheduleResults(ctx context.Context, signature, hintedNode, chosenNode string, otherNodes framework.SortedScoredNodes, cycleCount int64) {
+	logger := klog.FromContext(ctx)
+
+	startTime := time.Now()
+	defer func() {
+		metrics.FrameworkExtensionPointDuration.WithLabelValues(metrics.StoreScheduleResults, "", b.handle.ProfileName()).Observe(metrics.SinceInSeconds(startTime))
```

Comment:
I'm not sure if we should call StoreScheduleResults an extension point. Basically, the extension point is a collection of plugin runs. If you want to observe this latency, I feel like it's more appropriate to create another metric.